File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def solve(self, root, mx, mn):
        if root == None:
            return mx - mn
        mx = max(mx, root.val)
        mn = min(mn, root.val)

        return max(self.solve(root.left, mx, mn), self.solve(root.right, mx, mn))
        
    def maxAncestorDiff(self, root):
        return self.solve(root, root.val, root.val)
        

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
   
    int solve(TreeNode* root, int mx, int mn) {
        if(!root) return mx - mn;
        mx = max(mx, root->val);
        mn = min(mn, root->val);
        
        int x = solve(root->left, mx, mn);
        int y = solve(root->right, mx, mn);

        return max(x, y);
    }

    int maxAncestorDiff(TreeNode* root) {
        return solve(root, root->val, root->val);
    }
};